test/devicecontroller/DefaultDeviceController.test.ts

```diff
@@ -760,6 +760,18 @@ describe('DefaultDeviceController', () => {
       }
     });
 
+    it('Can chooseAudioInputDevice if pass in MediaDeviceInfo', async () => {
+      const spy = sinon.spy(navigator.mediaDevices, 'getUserMedia');
+      try {
+        await deviceController.chooseAudioInputDevice(
+          getMediaDeviceInfo('device-Id', 'audioinput', 'label', 'group-Id')
+        );
+      } catch (e) {
+        throw new Error('This line should not be reached.');
+      }
+      expect(spy.calledOnce).to.be.true;
```

Comment:
Nice test.

Could you add a spy on one of the functions that was involved in the recursion, too, and verify that it's only called once?

Reply:
getUserMedia is the call that happens in the recursion :)

Reply:
I meant one of our functions, like `chooseInputIntrinsicDevice`, which also appears in the recursion.

Reply:
Added `chooseInputIntrinsicDevice`